### DISCUSSION TITLE
cspy for ESPP with Ressource Constraints and Time Windows

### DISCUSSION THREAD

**Discussion by glanch** [torressa/cspy#117] [Q&A] [answered]
Dear community,
I am trying to solve a ESPP with ressource constraints and time windows (ESPPRCTW) as it occurs as a subproblem of a Branch-and-Price algorithm. I tried to directly solve the subproblem as a MIP, but it failed, probably due to the many Big-M constraints.

I can model the subproblem as an ESPPRCTW and hopefully use the cspy library for solving the subproblem, at least heuristically. 
My ESPRCTW contains both negative and positive edge weights. It may be the case that there are negative cycles. The travel time is independent of the edge weight and always positive.

Do you think solving this ESPPRCTW is doable with cspy? I already tried OR Tools but without any success as it seems that the routing solver cannot properly handle negative edge weights.

**Comment by torressa** [comment, marked as answer]
You can do this with the REF callbacks.
From Python, you can look at the example from the [vrpy/subproblem_cspy.py#L13](https://github.com/Kuifje02/vrpy/blob/master/vrpy/subproblem_cspy.py#L13)
